check_nightly_success/check-nightly-success/check.py

```diff
@@ -52,7 +53,24 @@ def main(
     now = datetime.now(tz=tz)
 
     latest_success = {}
+    # Rather frustratingly, the workflow runs returned from the GitHub API can
+    # have alternating ordering of `head_branch`
+    # e.g.
+    #   run[0]['head_branch'] == "branch-25.02"
+    #   run[1]['head_branch'] == "branch-25.04"
+    #   run[2]['head_branch'] == "branch-25.02"
+    #
+    # In this situation, the behavior of `itertools.groupby` is to only group
+    # _consecutive_ runs, so the results of the subsequent branch match (i.e.
+    # the second group of `branch-25.02` runs) will overwrite the results of the
+    # first one, potentially overwriting a previous success. The snippet below
+    # unifies the groups so it's more like a SQL groupby and there is no chance
+    # of overwriting.
+    branch_dict = defaultdict(list)
     for branch, branch_runs in itertools.groupby(runs, key=lambda r: r["head_branch"]):
+        branch_dict[branch] = itertools.chain(branch_dict[branch], list(branch_runs))
```

Comment:
I think this is equivalent to the following:
```suggestion
    for run in runs:
        branch_dict[run["head_branch"]].append(run)
```

Reply:
lol, you know when you've started with one paradigm and forget to look at the problem anew?